### ISSUE TITLE
Consider unsafely assuming that subclasses don't add `__await__`, for better `isawaitable` narrowing

### ISSUE THREAD

**Issue by alexmacnorthstar** [astral-sh/ty#1567]
### Summary

```
import asyncio
import inspect
from typing import Callable, Awaitable, Any

def foo_1():
  return [1,2,3]

async def foo_2():
  return [1,2,3]

async def test(
  foo: list[int] | Awaitable[list[int]]
):
  intlist = await foo if inspect.isawaitable(foo) else foo
  for x in intlist:
    print(x)

async def main():
  await test(foo_1())
  await test(foo_2())

asyncio.run(main())
```
Gives
```
error[not-iterable]: Object of type `object` is not iterable
  --> /Users/alex/Desktop/ty_test.py:15:12
   |
13 | ):
14 |   intlist = await foo if inspect.isawaitable(foo) else foo
15 |   for x in intlist:
   |            ^^^^^^^
16 |     print(x)
   |
info: It doesn't have an `__iter__` method or a `__getitem__` method
info: rule `not-iterable` is enabled by default

Found 1 diagnostic
```
The online playground shows that "intlist" is inferred as type "object" but it should be type "list[int]"

### Version

ty 0.0.1-alpha.26 (b225fd8b4 2025-11-10)

**Comment by carljm** [comment]
This is another case where our narrowing is strictly correct, but perhaps overly pedantic.

It is possible that there exists some user subclass of `list` which is awaitable and returns some other type (not even necessarily the same type as contained in the list, so not necessarily `int` in this case). So after checking `inspect.isawaitable(foo)`, we can't assume that the type of `foo` is `Awaitable[list[int]]` -- all we can assume is that it is some awaitable object, and awaiting it returns... something. That's why we infer `object` as the type of `intlist`.

One path forward here could be to place some restrictions on subclassing certain builtin types. Like, emit a diagnostic if you subclass `list` and add an `__await__` method to it. Then we could safely narrow in the expected way here. But maybe somebody out there will find this restriction unbearable?

Or we could just decide to make some unsafe assumptions in these cases.

**Comment by alexmacnorthstar** [comment]
I would assume that subclassing list would be the edge case and I'd actually want a warning if someone did that in our codebase - but the real world is a full of crazy stuff, and i could imagine that in a more complex codebase like numpy that deals with interesting container types that they might do those kind of gymnastics...

**Comment by sharkdp** [comment]
We discussed this internally, and we currently *do not* plan to make any change here.

This problem comes up with `isawaitable`, `iscallable`, but also `isinstance` checks on invariant container types like `isinstance(…, list)`. In the majority of cases, there are ways to restructure the code to make it correct. For example, for an `extract` function that takes `C | Awaitable[C]` and tries to get the `C` out, we can use `isinstance(obj, C)` instead of `isawaitable`:
```py
class C: ...

async def extract(obj: C | Awaitable[C]) -> C:
    if isinstance(obj, C):
        return obj
    else:
        return await obj
```
(https://play.ty.dev/cefc37cb-eaf5-48b5-8f33-3f768dcd0920)

Alternatively, we can mark the class `C` as `@final`:
```py
@final
class C: ...

async def extract(obj: C | Awaitable[C]) -> C:
    if isawaitable(obj):
        return await obj
    else:
        return obj
```
(https://play.ty.dev/f21074d6-8f39-4da3-a4f0-8feef8ab68a0)

The case described in the post here where `C` is replaced with `list[int]` is much more tricky. We can't use the first solution because `isinstance(obj, list)` has the same problem where we can't rule out that an awaitable might be a list subclass that would pass this check.  And the second solution obviously doesn't work because we can't make a builtin type `@final`.

Since there is an actual (at least theoretical) problem in the code here, I think it's fine to require a suppression comment for the implementation:
```py
async def extract(obj: list[int] | Awaitable[list[int]]) -> list[int]:
    if isawaitable(obj):
        # subclasses of `list` that are also awaitable will take this
        # branch, so `await obj` does not necessarily have to be of
        # type `list[int]`. We ignore this possibility here.
        return await obj  # ty: ignore
    else:
        return obj
```
(https://play.ty.dev/7bd1d2c9-1d03-4336-bcba-67e83b67ff1d)

A fourth solution is something that I've been thinking about recently — not sure if that's a good idea. Imagine `ty_extensions` had an `unsafe_narrowing` submodule that would provide variants of `isawaitable`, `iscallable`, `isinstance` etc. that would act exactly like their runtime equivalent, but provide an additional type parameter to control the (unsafe) narrowing behavior. Obviously it wouldn't make the code any safer, but it would let us write:
```py
from typing import Awaitable
from ty_extensions.unsafe_narrowing import isawaitable

async def extract(obj: list[int] | Awaitable[list[int]]) -> list[int]:
    if isawaitable[list[int]](obj):
        return await obj
    else:
        return obj
```
(Demo: https://play.ty.dev/c2cd6c78-c512-40a1-a188-564eac943812)

Another thing that we have discussed is the possibility to improve our diagnostics for cases like this and/or to make our `TypeIs` narrowing more "gradual". I have opened https://github.com/astral-sh/ty/issues/3375 to track this.

Otherwise, there is nothing actionable here, so I will close this ticket for now, but please feel free to comment here if you feel strongly that we should reconsider our decision.
